docs/source/format/CanonicalExtensions.rst

```diff
@@ -251,6 +251,27 @@ Variable shape tensor
    Values inside each **data** tensor element are stored in row-major/C-contiguous
    order according to the corresponding **shape**.
 
+.. _json_extension:
+
+JSON
+====
+
+* Extension name: `arrow.json`.
+
+* The storage type of this extension is ``StringArray`` or
+  or ``LargeStringArray`` or ``StringViewArray``.
+  Only UTF-8 encoded JSON is supported.
+
+* Extension type parameters:
+
+  This type does not have any parameters.
+
+* Description of the serialization:
+
+  Metadata is either an empty string or a JSON string with an empty object.
+  In the future, additional fields may be added, but they are not required
+  to interpret the array.
```

Comment:
Is there a specific reason we want to add this here for JSON, while we don't do this for uuid? Do we expect there to be metadata in the future?

(I would also argue that adding _optional_ metadata in the future should be regarded as backwards compatible, even if we don't specify that here now)

Reply:
An interesting metadata that could be put is ``{"schema": { ... insert here JSON schema .... }}`` to express contraints on the payload

Reply:
@tustvold suggested metadata could hold JSON schema which would not be the case with UUID which could justify the difference between the types.
I do agree with your reasoning that adding optional metadata later wouldn't be backwards incompatible, but perhaps we have users that need it now. @progger-dev any thoughts from you or someone on the bigquery team?

Reply:
> I would also argue that adding optional metadata in the future should be regarded as backwards compatible, even if we don't specify that here now

I agree, but in practice what is backwards compatible depends on what implementations do. So I think it's worth being extra clear here that it should not be an error if there is unexpected metadata.

Reply:
I had 3 things in my mind for what I would like to be able to use the metadata field for.

1. Parsing options that describe what features were used to generate the JSON. E.g. trailing commas allowed, unquoted field names, etc.
2. Schemas
3. Specialized metadata used internally (e.g. BigQuery might want to store what fields were columnarized, so we could reinstantiate it if the user exports data and then reimports it back).

None of these were requirements, but just some potential things that we might have used the metadata field for. I'm no longer at Google, so I'll tag @emkornfield for more recent thoughts from the BQ team.

Reply:
> 1. Parsing options that describe what features were used to generate the JSON. E.g. trailing commas allowed, unquoted field names, etc

ouch, I wouldn't call JSON with trailing commas, or unquoted field names JSON. Perhaps it would be wise to specify that JSON here is the one of https://www.json.org/json-en.html , that is ECMA-404 , otherwise that would defeat the purpose of interoperability between implementations

Reply:
I totally agree. I don't think we should canonicalize these non-standard extensions. I don't think producers should generate non-standard JSON. But I think consumers should be allowed to parse non-standard JSON. This is consistent with the spec in [RFC-8259](https://datatracker.ietf.org/doc/html/rfc8259#section-9).

This also begs the question of what we do when JSON parsing fails.

Reply:
Should we explicitly state in `CanonicalExtensions.rst` that this canonical extension type supports JSON _as defined in RFC-8259 (https://datatracker.ietf.org/doc/html/rfc8259)_?

Reply:
> 3\. Specialized metadata used internally (e.g. BigQuery might want to store what fields were columnarized, so we could reinstantiate it if the user exports data and then reimports it back).

Note that for such internal metadata, you are always free that any other key to the field metadata of the column, and you don't necessarily have to add that to the value of the `ARROW:extension:metadata` key

Reply:
I think I agree with Ian that stricter is better.  It's fine for different components within a system to agree they want to support a wider set of values.  However, once data leaves a solution, it should assume that most components will only be processing the to-spec JSON.  This document is primarily about interchange and so should be concerned with the latter case (data leaving a solution to be consumed by some unknown component).

Reply:
Added rfc8259 requirement (https://github.com/apache/arrow/pull/41257/commits/594945010e3b7d393b411aad971743ffcdbdbc8e).

To the original @jorisvandenbossche's question: it seems to me we're more in favour of adding optional metadata to JSON type than not so I'm leaving that language as is. Please correct me if I'm wrong.

Reply:
@progger-dev is 1. or 2. like what you have in mind?
1 . `Non-conformant or non-parsable JSON values should be considered invalid and halt producing or consuming process.`
2. `Non-conformant or non-parsable JSON values should be considered as null values by producers and parsers.`

Reply:
I was thinking about treating it the same as if you didn't support the extension type itself and just dealt with the underlying type and acts as a pass through.

But on second thought, this is not something for the specification as its likely going to be highly implementation dependent. Maybe its more of a recommendation?

Reply:
> I think we should also add language that specifies what happens when you receive non-conformant JSON or see any kind of parse errors. I think it's important for components that sit in between producers and consumers (and thus act as both).

Is it needed to specify anything about that? To me that seems a decision that an application that will actually use the JSON data will need or want to decide for themselves how to treat invalid input, while in many cases an Arrow implementation receiving such data might just be passing it along, and will typically not do any validation at all.